core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/helper/DefaultLocalDcHelper.java

```diff
@@ -0,0 +1,90 @@
+import com.datastax.oss.driver.internal.core.metadata.DefaultNode;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Optional;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DefaultLocalDcHelper extends BasicLocalDcHelper {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultLocalDcHelper.class);
+
+  public DefaultLocalDcHelper(
+      @NonNull InternalDriverContext context,
+      @NonNull DriverExecutionProfile profile,
+      @NonNull String logPrefix) {
+    super(context, profile, logPrefix);
+  }
+
+  /**
+   * This implementation fetches the user-supplied datacenter, if any, from the programmatic
+   * configuration API, or else, from the driver configuration. If no local datacenter is explicitly
+   * defined, this implementation will consider two distinct situations:
+   *
+   * <ol>
+   *   <li>If no explicit contact points were provided, this implementation will infer the local
+   *       datacenter from the implicit contact point (localhost).
+   *   <li>If explicit contact points were provided however, this implementation will throw {@link
+   *       IllegalStateException}.
+   * </ol>
+   *
+   * @return The local datacenter; always present.
+   * @throws IllegalStateException if the local datacenter could not be inferred.
+   */
+  @NonNull
+  @Override
+  public Optional<String> discoverLocalDc() {
+    Optional<String> optionalLocalDc = super.discoverLocalDc();
+    if (optionalLocalDc.isPresent()) {
+      return optionalLocalDc;
+    }
+    Set<DefaultNode> contactPoints = context.getMetadataManager().getContactPoints();
+    if (context.getMetadataManager().wasImplicitContactPoint()) {
+      // We only allow automatic inference of the local DC in this specific case
+      assert contactPoints.size() == 1;
+      Node contactPoint = contactPoints.iterator().next();
+      String localDc = contactPoint.getDatacenter();
+      if (localDc != null) {
+        LOG.debug(
+            "[{}] Local DC set from implicit contact point {}: {}",
+            logPrefix,
+            contactPoint,
+            localDc);
+        return Optional.of(localDc);
+      } else {
+        throw new IllegalStateException(
+            "The local DC could not be inferred from implicit contact point, please set it explicitly (see "
+                + DefaultDriverOption.LOAD_BALANCING_LOCAL_DATACENTER.getPath()
+                + " in the config, or set it programmatically with SessionBuilder.withLocalDatacenter)");
+      }
+    } else {
+      throw new IllegalStateException(
+          "Since you provided explicit contact points, the local DC must be explicitly set (see "
+              + DefaultDriverOption.LOAD_BALANCING_LOCAL_DATACENTER.getPath()
+              + " in the config, or set it programmatically with SessionBuilder.withLocalDatacenter). "
+              + "Current contact points are: "
```

Comment:
We should also include the available DCs.

Reply:
oh man... I see that `formatNodes()` already include the DC name of each contact point, thanks Alex for pointing that out.

nvm!